### DISCUSSION TITLE
Metadata Operator for Inline header math

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#296] [Design & Development]
I want to propose the following base class design for an inline Metadata editor.  Peruse this and jump past the code to see the suggestion on how it could be used as a base class:
```
class MetadataOperator(ABC):
    """
    Base class for a set of inline Metadata editors. That is, there 
    are many instances where Metadata attributes need to be altered 
    during a workflow where it is either unnecessary or inappropriate 
    to access the database.   e.g. we have found updates in MongoDB
    can be very slow if done one transaction at a time so it can 
    streamline processing to edit metadata on the fly.   This base 
    class defines the API for a suite of tools for editing metadata 
    on the fly. 
    """
    @abstractmethod
    def apply(self,d):
        """
        Implementations are required to implement this method.  It must
        accept a MsPASS data object, d, (TimeSeries, Seismogram, TimeSeriesEnsemble,
        or SeimogramEnsemble) and apply the operator the implementation defines 
        to the Metadata of d.   It should throw a MsPASSError exception if 
        d is not one of the expected data types.  It should, on the other 
        hand, handle d==None gracefully and do nothing if d is None.  
        
        This method must return an (potentially but not guarnateed edited)
        version of d.  That is essential to allow this method to be used 
        as the function in a parallel map operator. 
        """
        pass
    def input_is_valid(self,d):
        """
        This is a common method all instances of this class needed for
        mspass.  It uses isinstance tests of d to standardize the test that
        the input is valid data.  It returns True if d is one a valid data
        object known to mspass.  Returns false it not.  Caller must decide
        what to do if the function returns false.
        """
        if isinstance(d,(TimeSeries,Seismogram,TimeSeriesEnsemble,SeismogramEnsemble)):
            return True
        else:
            return False
    @abstractmethod
    def check_keys(*arg,**karg):
        """
        All implementation should implement this checker even if all it contains 
        is a pass.  It should validate the keys are defined in data to be 
        handled in the apply method.   An early call in apply should always 
        be to call this method.  
        """
        pass
    
    @abstractmethod
    def check_operation(*arg,**karg):
        """
        All implementations should implement this method even if they choose 
        to ignore it.   It should be used to guarantee the operator the class 
        defines will succeed on the datum sent to the apply method.   
        e.g. any standard arithmetic operations will throw a TypeError if 
        one of the operands is a string.   This method should be used to 
        make the operator as bombproof as possible logging a message 
        to the datum rather than aborting if there is an issue.  Some 
        classes may want to implement this as pass because it makes 
        no sense - e.g. setting a constant value.   Those are the 
        exception, however, so the api dogmatically demands these 
        be implemented even if they do nothing.
        """
        pass
        

    def log_edit(self,d,testname,message,severity=ErrorSeverity.Informational):
        """
        This base class method is used to standardize the error logging 
        functionality of all editors.   It writes a standardized 
        message to simplify writing of subclasses - they need only 
        define the testname (normally the name of the subclass) and 
        format a specific message to be posted.  
        
        Note most subclasses will may want to include a verbose option 
        (or the reciprocal silent) and only write log messages when 
        verbose is set true.  
        
        :param d:  MsPASS data object to which elog message is to be 
          written.
        :param testname: is the string assigned to the "algorithm" field 
        of the message posted to d.elog.
        :param message:  specialized message to post - this string is added 
        to an internal generic message.
        :param severity:  ErrorSeverity to assign to elog message 
        (See ErrorLogger docstring).  Default is Informational
        """
        if self.input_is_valid(d):
            edit_message = "Altered by apply method of this class.  Details:\n"
            edit_message += message
            d.elog.log_error(testname,edit_message,severity)
        else:
            raise MsPASSError("MetadataOperator.apply method received invalid data;  must be a MsPASS data object",
                              ErrorSeverity.Fatal)
```
Noting this passes spyder's syntax checks but may have some issues that are irrelevant right now. 

The key idea is that would be a base class for a set of Metadata operators.  That is, we need the equivalent of what seismic unix uses to set and change header data.   Let a, b, and d be three (numeric but not necessarily restricted to that) Metadata attributes accessible by keys "keya",  "keyb", and "keyd" respectively.   Also let c be any constant.   Some basic operators that would cover a lot of options are:
a += b or d = a + b
a *= b or d = a*b
a /= b or d = a/b
a -= b or d = a - b
a *= c or d = a*c
a +- c or d = a+c
a /= c or d = a/c
a -= c or d = a - c

there are so good reasons for mod operators like d = a%b or d=a%c or the remainder version of same.

There is a need to rename the keys and set all values to a constant.

I think this collection of subclasses of the above base class could do this:
```
class ChangeKey(MetadataOperator):
class SetValue(MetadataOperator):  
# all arithmetic operators need an output key to place result that can be the same as one of the inputs
# These are unary - operate using a constant value set on constructoin
class Add(MetadataOperator):
class Multiply(MetadataOperator):
class Subtract(MetadataOperator):
class Divide(MetadataOperator):
class Remainder(MetadataOperator):
class Mod(MetadataOperator):
# These are binary operators using two keys set in constructor
class Add2(MetadataOperator):
class Multiply2(MetadataOperator):
class Subtract2(MetadataOperator):
class Divide2(MetadataOperator):
class Remainder2(MetadataOperator):
class Mod2(MetadataOperator):
```
Where the `2` postfix in all cases means it would implement the binary version.  e.g. `Add` would implement a +=c or d=a+x (the same function could handle both cases automatically) while `Add2` would implement a+=b or d += a + b.   

ChangeKey would change the key used to access an attribute or optionally copy it to a new key.  
SetValue would be used to set a key to a constant value for all data for which it would be called.  

The classes would be set up with constructors that initialize keys and c (when relevant).  e.g. this would be how you would use 'Multiple` to convert from units of meter to km on attribute `dnorth` setting the result with the (made up ) key `dnkm`:
```
import mspasspy.edit as MDO
operator = MDO.Multiply('dnorth','dnkm',0.001)
... code to create a bag/rdd of mspass data objects.  Call that container mydata ...
mydata.map(operator.apply)
```
Finally, this probably needs a chain method that would cascade mulitple instances of MetadataOperator subclasses to be applied sequentially.  One could build up quite complex calculations that way.

Some possible extensions:
1.  Much of this could probably be implemented for any Metadata attribute value for which the relevant operators applied.  e.g. + is a standard operator with a well defined meaning for strings while * is not.  
2. I think it might be possible to implement `class Function(MetadataOperator)` wherein the constructor could load a numeric function like math.log, math.log10, math.exp, etc.  Because python handles function options like any other class I don't see why this couldn't work if you used a constructor `def __init__(self,function):`  where function would be something like `math.exp`.  

What do you guys think of this approach?  It is similar to what I just implemented for kill operators and should probably live the the same module.  Also this uses relatively simple naming for the subclass names, but maybe there is some more clever pythonic way to overload the operator symbols like +=, -=, etc.   I don't really think with this paradigm but there might be a completely different way to implement the concept I'm trying to do here:  apply operators to Metadata in map operations to do inline editing and header math.

**Comment by Yangzhengtang** [comment]
This approach looks good to me. However, we might want the [history_manager](https://github.com/mspass-team/mspass/blob/b18f40f46c3134fcc782df84423605abfd15e5ea/python/mspasspy/global_history/manager.py#L376) to capture the operator and its paramenters. The example discribed above (`_mydata.map(operator.apply)_`) would hide those information. So maybe we need to add a wrapper function:
```python
def operateOnMetaData(d, op, *arg, **karg):
    '''
    :param d: MsPass data object to operate on.
    :param op: Name of the operator, for example "Add" or "Subtract"
    :param *arg and **karg: Parameters used to construct a MetadataOperator
    '''

    if op == "Add":
        operator = MDO.Add(*arg, **karg)
    elif op == "Subtract":
        operator = MDO.Subtract(*arg, **karg)
    elif op == "Multiply":
        operator = MDO.Multiply(*arg, **karg)
    '''
    ... (Check if it matches other opertors)
    '''
    else:
        raise MsPASSError("Undefined op name.")

    return operator.apply(d)
```
Then the usage would be `mydata.map(operateOnMetaData, 'Multiply', 'dnorth','dnkm',0.001)`. And the parameters will be captured by history_manager.

(The operators can be overloaded in python. This [link](https://www.programiz.com/python-programming/operator-overloading) and [emulating-numeric-types](https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types) show how to do the overloading. Maybe it is another feasible approach?)

**Reply by pavlis** [reply]
Very good point about history.  That didn't cross my mind and it should have.  I think my Executioner class has the same issue.   I think in both cases history is more appropriately enabled in the "apply" method ("kill_if_true" for the Executioners).   i.e. the abstract base class should require a "save_history" (or whatever our standard name is - I'll need to check) argument.   What you suggest wouldn't work with composite operators that apply a chain of operations.  In the subclasses of Executioner I called that "FiringSquad".  I didn't suggest a name for this set of operators.   

Need to think about this, but I think it might work best to enable history in the constructor and add a base class method to turn history on and off.  Let me play with that idea.  

Your wrapper might be a good idea in any case as a simplified interface.  It's dark side, however, is the operator has be constructed for each call.   This is a tiny class, however, so that is probably a minor efficiency issue.  Let me also play with that idea further today

As to the use of operators that was helpful.  However, when I think it through I think it would be very confusing to users for the same reason we depricated += in the ccore bindings of Metadata.  Recall we found if you had to TimeSeries objects d1 and d2 and used a construct like `d1 += d2` the result was ambiguous.  Does that mean (in C++ syntax)  Metadata::operator+= or CoreTimeSeries::operator+=.    This is one of those type examples of the dark side of multiple inheritance in OOP.  The implementation of that approach, furthermore, would have to be done completely differently.   I also can't think of a way you could set keys for Metadata through that mechanism.   So, unless you see a way to do that cleanly I'm going to start messing with the above implementation including adding a history capability.

**Comment by pavlis** [comment]
I wonder if one of you could help me out on this one.   I'm a bit shaky on how decorators work and was trying to use the mspass_func_wrapper in a class declaration to handle history (as noted above) as follows:
```
class MetadataGT(Executioner):
    @mspass_func_wrapper
    def kill_if_true(self, d, *args,
          object_history=False, alg_name=None, alg_id=None, dryrun=False,
          inplace_return=True, function_return_key=None, **kwargs):
```
Noting for your reference this is the signature of mspass_func_wrapper:
```
def mspass_func_wrapper(func, data, *args, object_history=False, alg_id=None, alg_name=None, dryrun=False,
                        inplace_return=False, function_return_key=None, **kwargs):
```
The decorator throws a TypeError exception when used this way.  The reason is clear when I step through this (incorrect) implementation.  When the decorator is called in this context it receives `self` instead of the `data` arg it expect as arg0 of the wrapped method function.   

Do any of you know a simple way to fix this short of writing a new wrapper for class methods that would be painfully similar to mspass_func_wrapper?  I've not been successful with google searches, although there are a lot of potential sources and I've not dug deep enough.  If I understood decorators better the solution might be obvious without leaning on internet searches.

**Comment by wangyinz** [comment]
Our decorator was designed for functions, so it is not expected to handle methods like this. I guess the best way to do it is to create a new decorator, maybe called `mspass_method_wrapper`. The initial implementation could just be changing the parameters of `mspass_func_wrapper` and keep everything else the same. However, we might want to consider what to do with the additional `self` parameter that could potentially add more info to the history.

**Comment by pavlis** [comment]
I realized I think we should also implement a function with this signature:
```
def edit_readonly(data_object,names_to_change=[],log_changes=False):
```
The idea of this function is a simplified way to rename attributes pushed to mongodb with names like `READONLYERROR_sta, READONLYERROR_net, and READONLYERROR_channel_lat` to the standard tags `sta, net, and lat` respectively (for the example) .  

The key parameter is `names_to_change`.   I suggest it should be a list of metadata data keys that would be passed through the ChangeKey header editor that is the subject of this discussion page.  For the example above one would use: `names_to_change=["sta","net","channel_lat"]`.   

My prototype of this stuff already has a stock implementation of log_changes.  If false edits would be preformed silently but if true all changes would be logged to elog.   That should normally be off as such changes could get huge for some workflows.

**Comment by pavlis** [comment]
Another extension we might consider is a more extensive changer than the READONLYERROR wrapper noted above.  We might want to design a header mapping function that would change namespaces to impose a standard alias list. e.g. we have previously discussed things like a standard set of SAC or SEGY aliases.
